tests/utils.go

```diff
@@ -646,6 +646,32 @@ func NewRandomMigrationForVm(vm *v1.VirtualMachine) *v1.Migration {
 	return v1.NewMinimalMigrationWithNS(ns, vm.ObjectMeta.Name+"migrate"+rand.String(5), vm.ObjectMeta.Name)
 }
 
+func NewRandomVMWithWatchdog() *v1.VirtualMachine {
+	vm := NewRandomVMWithDirectLun(2, false)
+	vm.Spec.Domain.Devices.Serials = []v1.Serial{
+		{
+			Type: "pty",
+			Target: &v1.SerialTarget{
+				Port: newUInt(0),
+			},
+		},
+	}
+	vm.Spec.Domain.Devices.Consoles = []v1.Console{
+		{
+			Type: "pty",
+			Target: &v1.ConsoleTarget{
+				Type: newString("serial"),
+				Port: newUInt(0),
+			},
+		},
+	}
+	vm.Spec.Domain.Devices.Watchdog = &v1.Watchdog{
+		Model:  "i6300esb",
+		Action: "poweroff",
```

Comment:
Is "poweroff" interpreted by libvirt, or how do we know that the poweroff should lead to "failed" instead of "succeeded"?

Reply:
the "right way" to handle stuck guest is using the pvpanic device. I do not believe anyone bothers with watchdog support anymore. 
Watchdog is ok from completeness perspective, but I do not expect anyone really using that in the real world

Reply:
Can it be that we want both? pvpanic seems to be focused around kernel panics, whereas the watchdog device can cover a brader range of scenarios where a guest is "stuck". Can that be?

Reply:
it's cooperative anyway, and the supported base seems to be smaller than with pvpanic already.
So, yeah, both are good to have so you have more options, but I expect "users" to use pvpanic. It gets more focus by qemu/libvirt developers as well ("panic" device)

Reply:
We really should strive to deprecate some of the unused devices and not support them from the beginning. Any watchdog device apart from the pvpanic for example.

Reply:
> Is "poweroff" interpreted by libvirt, or how do we know that the poweroff should lead to "failed" instead of "succeeded"?

power off in this case results in libvirt reporting the domain as crashed.

Reply:
> Watchdog is ok from completeness perspective, but I do not expect anyone really using that in the real world

I implemented this feature because I'm considering using it. pvpanic only handles a small subset of what the watchdog functionality can detect.

> We really should strive to deprecate some of the unused devices and not support them from the beginning. Any watchdog device apart from the pvpanic for example.

@mykaul What do you recommend to use if I want more advanced monitoring outside of just detecting the kernel paniced?  Lets say for example the VM is a k8s node, and I want the VM to be shutdown if the kubelet process is unrecoverable for some reason.

Reply:
App-level monitoring - there's really no way around it if you want to ensure the App is running (and if not, restart the hosting VM).

Reply:
> App-level monitoring - there's really no way around it if you want to ensure the App is running (and if not, restart the hosting VM).

right. This describes eactly what I'm interested in using the watchdog daemon for.

Reply:
Perhaps the qemu guest agent should be extended (Something like https://github.com/guillon/qemu-plugins/tree/master/scripts/qemu-guest-agent ).
The watchdog device (apart from pvpanic) simply not supported in new OS, so it won't help us.
(As an alternative, any app-monitoring using virtio-serial can be set up, I reckon)

Reply:
> The watchdog device (apart from pvpanic) simply not supported in new OS, so it won't help us.

@mykaul I don't believe this is the case.

Watchdog usage appears fairly common and supported in most distros. Alpine linux's virt image (which is super small) even chose to include the watchdog daemon by default. They are extremely space conscious for this image. Cloud-init for example isn't even included in this distro, but watchdog is.

Reply:
@mykaul is it going to be removed from the kernel of newer distributions, or what do you mean by not supported?

Reply:
I was referring to Windows mainly (where it does NOT come inbox, for example).
I would assume at some point it'd be replaced with pvpanic, as the HW itself is getting harder to get anyway...

Reply:
> I was referring to Windows mainly (where it does NOT come inbox, for example). I would assume at some point it'd be replaced with pvpanic, as the HW itself is getting harder to get anyway...

@mykaul  In the future, please clearly state the details pertaining to why you're asserting a statement like...

"We really should strive to deprecate some of the unused devices and not support them from the beginning. Any watchdog device apart from the pvpanic for example."

and

"The watchdog device (apart from pvpanic) simply not supported in new OS, so it won't help us."

Windows is an important use case to us, but it is not "the" use case. It would have saved us a bit of time to know you're talking primiarily about Windows from the beginning.

Understand that when you making sweeping statements like this without providing the relevant information up front that it involves other peoples time. It involves me researching why people are migrating away from watchdog, which I found nothing... It wastes community meeting time asking if anyone has heard about why watchdog is being depreciated, which no one has any information about (because it's not happening). Then we have to continue to draw out the details about what exactly you mean by your statements because it is at odds with reality (minus Windows apparently)

I definitely appreciate the feedback. We just need to be more efficient in how we communicate.

Reply:
Let me clarify the 'deprecate some of the unused devices' - the wealth of devices we have today (and it's growing) is a pain to support, test and certify. We should strive to limit them. I am under the impression that specifically that one can be replaced with pvpanic and therefore I'd use this opportunity and NOT support it from the get go (same for Cirrus for video device, btw, rtl8139 for network, IDE for CDROM, ...)

Reply:
@mykaul:
> I am under the impression that specifically that one can be replaced with pvpanic [...]

As @davidvossel mentioned above, pvpanic only handles a small subset of what the watchdog device can detect.

Reply:
@ehabkost can you clarify the subset?And while you're at it perhaps a more information about future plans?:) .
How about support (for either) on non-x86 platforms? There is indeed not much information to be found other than relatively old discussion threads

Reply:
@michalskrivanek 

https://linux.die.net/man/8/watchdog and https://linux.die.net/man/5/watchdog.conf. The watchdog daemon is capable of much more than detecting kernel panics. Those man pages enumerate all the tests watchdog is capable of performing on the guest.

Those tests can be configured using cloud init to write to the watchdog config. Essentially this lets someone create a custom health monitoring policy they can assign to a VM. For example, if someone wants to reset the VM when free memory hits a certain threshold, watchdog can do that and pvpanic can not.

Reply:
@michalskrivanek , afaik, pvpanic will only react to guest panic events, whereas watchdog monitors whether the guest OS is stuck (even if there is no free memory to operate). For the watchdog to work it needs a watchdog daemon to run in the guest. Libivrt then can take action.
Afaik, watchdog doesn't act on guest panic events, which is why pvpanic support is also needed to cover the whole scenario.

Reply:
watchdog should reacts to panic events as well. If the kernel panics, the watchdog daemon in the guest no longer functions which will result in the watchdog timer expiring and the VM getting powered off.